azurerm/internal/services/resource/resource_arm_provider_registration.go

```diff
@@ -0,0 +1,188 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/2017-03-09/resources/mgmt/resources"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmResourceProviderRegistration() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmResourceProviderCreateUpdate,
+		Read:   resourceArmResourceProviderRead,
+		Update: resourceArmResourceProviderCreateUpdate,
+		Delete: resourceArmResourceProviderDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"resource_provider_namespace": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceArmResourceProviderCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Resource.ProvidersClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	resourceProviderNamespace := d.Get("resource_provider_namespace").(string)
+	expand := ""
+
```

Comment:
we'll also probably need to add a check to ensure that automatic Resource Provider registration is disabled here - since otherwise Terraform and this resource provider will conflict, which isn't ideal

that information isn't exposed outside the initial provider block right now - so that may need to be threaded through [from here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/clients/builder.go#L20) into [the account](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/clients/builder.go#L62) where it should then be available as:

```
meta.(*clients.Client).Account.SkipResourceProviderRegistration
```

Reply:
@tombuildsstuff, this makes sense but what about the scenario/situation where we are not automatically registering a resource and the automatic registration is not disabled?

Reply:
@grayzu good question - in practice we'll continue to add to the automatic RP registration list over time as new services appear (whether this is defaulted to registering them by default/not is a separate question) - so it's probably safer to check this regardless, rather than returning an error later should an RP get added to the "auto register" list?